File: scripts/import-logos-zip.mjs
/**
 * Extracts logos from a zip where filenames = club names,
 * fuzzy-matches each to a club in the DB, stores as <id>.<ext>
 * in artifacts/api-server/logos/, and updates image_url.
 * Replaces any existing logo for a matched club.
 */
import mysql from "mysql2/promise";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { execSync } from "child_process";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const LOGOS_DIR  = path.resolve(__dirname, "../artifacts/api-server/logos");
const ZIP_PATH   = path.resolve(__dirname, "../attached_assets/Club_logos_1_1779295243345.zip");
const EXTRACT_TMP = "/tmp/club_logos_extracted";

// ── 1. Extract zip via python3 (built-in zipfile) ───────────────────────────
if (fs.existsSync(EXTRACT_TMP)) execSync(`rm -rf "${EXTRACT_TMP}"`);
fs.mkdirSync(EXTRACT_TMP, { recursive: true });

execSync(
  `python3 -c "import zipfile,sys; zipfile.ZipFile(sys.argv[1]).extractall(sys.argv[2])" "${ZIP_PATH}" "${EXTRACT_TMP}"`,
  { stdio: "inherit" }
);

const extractedFiles = fs.readdirSync(EXTRACT_TMP);
console.log(`Extracted ${extractedFiles.length} files\n`);

// ── 2. Connect + fetch all clubs ─────────────────────────────────────────────
const conn = await mysql.createConnection({
  host: process.env.MYSQL_HOST,
  port: parseInt(process.env.MYSQL_PORT || "3306"),
  user: process.env.MYSQL_USER,
  password: process.env.MYSQL_PASSWORD,
  database: process.env.MYSQL_DATABASE,
});

const [clubs] = await conn.query("SELECT id, name FROM clubs ORDER BY id");
console.log(`Loaded ${clubs.length} clubs from DB\n`);

// ── 3. Fuzzy match helper ────────────────────────────────────────────────────
function normalise(s) {
  return s.toLowerCase().replace(/[^a-z0-9 ]/g, " ").replace(/\s+/g, " ").trim();
}

function similarity(a, b) {
  // Dice-coefficient on bigrams
  const bigrams = s => {
    const set = new Map();
    for (let i = 0; i < s.length - 1; i++) {
      const bg = s.slice(i, i + 2);
      set.set(bg, (set.get(bg) || 0) + 1);
    }
    return set;
  };
  const biA = bigrams(a), biB = bigrams(b);
  let intersection = 0;
  for (const [bg, cnt] of biA) {
    if (biB.has(bg)) intersection += Math.min(cnt, biB.get(bg));
  }
  const total = a.length - 1 + b.length - 1;
  return total === 0 ? 0 : (2 * intersection) / total;
}

const normClubs = clubs.map(c => ({ id: c.id, name: c.name, norm: normalise(c.name) }));

function bestMatch(filename) {
  const q = normalise(filename);
  let best = { score: 0, club: null };
  for (const c of normClubs) {
    const score = similarity(q, c.norm);
    if (score > best.score) best = { score, club: c };
  }
  return best;
}

// ── 4. Process each logo ─────────────────────────────────────────────────────
fs.mkdirSync(LOGOS_DIR, { recursive: true });
const updates = [];
const lowConfidence = [];

for (const filename of extractedFiles) {
  const parsed = path.parse(filename);
  const stem = parsed.name;
  let ext = parsed.ext.toLowerCase();
  if (ext === ".jfif") ext = ".jpg";
  if (ext === ".PNG") ext = ".png";

  const { score, club } = bestMatch(stem);
  if (!club) continue;

  // Remove any existing logo for this club ID
  const existing = fs.readdirSync(LOGOS_DIR).filter(f => f.match(new RegExp(`^${club.id}\\.`, "i")));
  for (const old of existing) fs.unlinkSync(path.join(LOGOS_DIR, old));

  // Copy new logo
  const dest = path.join(LOGOS_DIR, `${club.id}${ext}`);
  fs.copyFileSync(path.join(EXTRACT_TMP, filename), dest);

  const url = `/api/logos/${club.id}${ext}`;
  updates.push({ url, id: club.id });

  const mark = score >= 0.75 ? "✅" : "⚠️ ";
  console.log(`${mark} [${score.toFixed(2)}]  "${stem}"  →  [${club.id}] "${club.name}"  →  ${url}`);
  if (score < 0.75) lowConfidence.push({ stem, matched: club.name, score });
}

// ── 5. Batch update DB ───────────────────────────────────────────────────────
if (updates.length > 0) {
  const caseExpr = updates.map(u => `WHEN ${u.id} THEN '${u.url}'`).join(" ");
  const ids = updates.map(u => u.id).join(",");
  const [result] = await conn.query(
    `UPDATE clubs SET image_url = CASE id ${caseExpr} END WHERE id IN (${ids})`
  );
  console.log(`\n✅ Updated ${result.affectedRows} clubs in the database`);
}

await conn.end();

// ── 6. Summary ────────────────────────────────────────────────────────────────
const [[{ total }]] = await (async () => {
  const c2 = await mysql.createConnection({
    host: process.env.MYSQL_HOST, port: parseInt(process.env.MYSQL_PORT || "3306"),
    user: process.env.MYSQL_USER, password: process.env.MYSQL_PASSWORD,
    database: process.env.MYSQL_DATABASE,
  });
  const r = await c2.query("SELECT COUNT(*) as total FROM clubs WHERE image_url IS NOT NULL AND image_url != ''");
  await c2.end();
  return r;
})();

console.log(`\nTotal logos in DB now : ${total}`);
console.log(`Logos processed        : ${updates.length}`);
console.log(`Low-confidence matches : ${lowConfidence.length}`);
if (lowConfidence.length) {
  console.log("\nReview these:");
  for (const { stem, matched, score } of lowConfidence) {
    console.log(`  [${score.toFixed(2)}] "${stem}"  →  "${matched}"`);
  }
}

File: scripts/import-logos-zip.py
"""
Extracts logos from a zip where filenames = club names,
fuzzy-matches each to a club in the DB, stores as <id>.<ext> in
artifacts/api-server/logos/, and updates image_url.
Replaces any existing logo for a matched club.
"""
import os, sys, zipfile, shutil, difflib, re
from pathlib import Path
import mysql.connector

LOGOS_DIR = Path(__file__).parent.parent / "artifacts/api-server/logos"
ZIP_PATH  = Path(__file__).parent.parent / "attached_assets/Club_logos_1_1779295243345.zip"
EXTRACT_TMP = Path("/tmp/club_logos_extracted")

# ── 1. Extract zip ────────────────────────────────────────────────────────────
if EXTRACT_TMP.exists():
    shutil.rmtree(EXTRACT_TMP)
EXTRACT_TMP.mkdir(parents=True)

with zipfile.ZipFile(ZIP_PATH) as z:
    z.extractall(EXTRACT_TMP)

extracted = list(EXTRACT_TMP.iterdir())
print(f"Extracted {len(extracted)} files")

# ── 2. Connect to DB and fetch all clubs ─────────────────────────────────────
conn = mysql.connector.connect(
    host=os.environ["MYSQL_HOST"],
    port=int(os.environ.get("MYSQL_PORT", 3306)),
    user=os.environ["MYSQL_USER"],
    password=os.environ["MYSQL_PASSWORD"],
    database=os.environ["MYSQL_DATABASE"],
)
cur = conn.cursor(dictionary=True)
cur.execute("SELECT id, name FROM clubs ORDER BY id")
clubs = cur.fetchall()
print(f"Loaded {len(clubs)} clubs from DB")

club_names = [c["name"] for c in clubs]
club_by_name = {c["name"]: c["id"] for c in clubs}

def normalise(s):
    """lowercase, strip punctuation/extra spaces for comparison"""
    s = s.lower()
    s = re.sub(r"[^a-z0-9 ]", " ", s)
    return re.sub(r"\s+", " ", s).strip()

norm_club_names = [(normalise(n), n) for n in club_names]

def best_match(query_name):
    q = normalise(query_name)
    scored = [(difflib.SequenceMatcher(None, q, nc).ratio(), orig) for nc, orig in norm_club_names]
    scored.sort(reverse=True)
    return scored[0]  # (score, club_name)

# ── 3. Match and copy ─────────────────────────────────────────────────────────
LOGOS_DIR.mkdir(parents=True, exist_ok=True)
updates = []  # (image_url, club_id)
unmatched = []

for img_path in extracted:
    stem = img_path.stem          # filename without extension = club name
    ext  = img_path.suffix.lower()
    if ext == ".jfif":
        ext = ".jpg"              # normalise jfif → jpg

    score, matched_club = best_match(stem)
    club_id = club_by_name[matched_club]

    # Remove any existing logo file for this club (any extension)
    for old in LOGOS_DIR.glob(f"{club_id}.*"):
        old.unlink()

    dest = LOGOS_DIR / f"{club_id}{ext}"
    shutil.copy2(img_path, dest)

    url = f"/api/logos/{club_id}{ext}"
    updates.append((url, club_id))

    marker = "✅" if score >= 0.75 else "⚠️ "
    print(f"  {marker} [{score:.2f}]  '{stem}'  →  [{club_id}] '{matched_club}'  →  {url}")
    if score < 0.75:
        unmatched.append((stem, matched_club, score))

# ── 4. Batch-update DB ───────────────────────────────────────────────────────
if updates:
    case_sql = " ".join(f"WHEN {cid} THEN %s" for _, cid in updates)
    id_list  = ",".join(str(cid) for _, cid in updates)
    values   = [url for url, _ in updates]
    cur.execute(f"UPDATE clubs SET image_url = CASE id {case_sql} END WHERE id IN ({id_list})", values)
    conn.commit()
    print(f"\n✅ Updated {cur.rowcount} clubs in the database")

cur.close()
conn.close()

# ── 5. Summary ────────────────────────────────────────────────────────────────
print(f"\nTotal logos processed : {len(updates)}")
print(f"Low-confidence matches: {len(unmatched)}")
if unmatched:
    print("\nReview these low-confidence matches:")
    for stem, matched, score in unmatched:
        print(f"  [{score:.2f}] '{stem}'  →  '{matched}'")
